File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.js
/**
 * @param {number[][]} questions
 * @return {number}
 */
var mostPoints = function(questions) {
    
    function util(index,dp){
        if(index>=questions.length) return 0;
        
        if(dp[index]!=-1) return dp[index];
        
        var take = questions[index][0] + util(index+questions[index][1]+1,dp);
        var leave = util(index+1,dp);
        
        return dp[index] = Math.max(take,leave);
    }
    
    var dp = new Array(questions.length).fill(-1);
    return util(0,dp);
};

File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.cpp
class Solution {
public:
    
    long long util(vector<vector<int>>& questions, int index,vector<long>& dp){
        
        if(index>=questions.size()) return 0;
        
        if(dp[index]!=-1) return dp[index];
        
        long long take = questions[index][0] + util(questions,index+questions[index][1]+1,dp);
        long long leave = util(questions,index+1,dp);
        
        return dp[index] = max(take,leave);
        
    }
    
    long long mostPoints(vector<vector<int>>& questions) {
       
        vector<long> dp(questions.size(),-1);
        return util(questions,0,dp);
        
    }
};